### PR TITLE
[5.8] Add support for app->call() using a sequential array as parameters

### PR DESCRIPTION
**Pull request by imanghafoori1**
- Trying to fix this issue :
https://github.com/laravel/framework/issues/26851
https://github.com/laravel/framework/pull/27751

- This is a re submission of #26920 but in a way that avoids tampering with the current code as much as possible. Most of the added code is isolated within two new protected methods.

- I can not imagine more tests to proof this PR

- **Side benefit:**  If this gets added to the framework, we can use `$container->call(` in the base facade class to call the intended method. (I will make the subsequent PR then)
currently is it called like this :
https://github.com/laravel/framework/blob/00624592ef044c2ddae5cbc324d39928f5b4bb70/src/Illuminate/Support/Facades/Facade.php#L237

```php
         // can be changed to :
        return static::$app->call($method, $args);

```
(Since the `$args` variable is a sequential array of parameters)

So the end users can enjoy `method injection` on POPOs when they use facades to call methods on classes.

```php
// we use a real-time facade to call MyClass::myMethod
\Facades\MyClass::myMethod();
````

```php
// the POPO which needs method injection
class MyClass {
    public function myMethod(SomeClass $obj) {
        // I have my object injected here...
    }
}
```

- About BoundMethodTest :
This style of unit testing a class ensures that if it works correctly with one type of $callback (e.g closures), it will also work well with other types.

So we need to only test it against Closures and be sure `class@method` notation is also ok.

### CODE REVIEW

File: src/Illuminate/Container/BoundMethod.php [hunk 1]
```diff
@@ -170,6 +175,36 @@ protected static function addDependencyForCallParameter($container, $parameter,
         }
     }
 
+    /**
+     * Get the dependency for the given call parameter.
+     *
+     * @param  \Illuminate\Container\Container $container
+     * @param $callback
```
Comment (setkyar) [inline src/Illuminate/Container/BoundMethod.php]: Add `$callback` type?

File: src/Illuminate/Container/BoundMethod.php [hunk 2]
```diff
@@ -180,4 +215,35 @@ protected static function isCallableWithAtSign($callback)
     {
         return is_string($callback) && strpos($callback, '@') !== false;
     }
+
+    /**
+     * @param $container
```
Comment (setkyar) [inline src/Illuminate/Container/BoundMethod.php]: Add `$container` type?

Comment (imanghafoori1) [reply src/Illuminate/Container/BoundMethod.php]: Will fix them, thanks.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**